File: 2016/day07_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any

type IPV7 = tuple[list[str], list[str]]


def solution(content: list[str]) -> Any:
    return len(
        [ip for ip in ((parse(line.strip())) for line in content) if support_tls(ip)]
    )


def support_tls(ip: IPV7) -> bool:
    net, hnet = ip
    return any(has_abba(part) for part in net) and not any(
        has_abba(part) for part in hnet
    )


def has_abba(part: str) -> bool:
    if len(part) < 4:
        return False
    return any(
        a == d and b == c and a != b
        for a, b, c, d in zip(part, part[1:], part[2:], part[3:])
    )


def parse(line: str) -> IPV7:
    nets: list[str] = []
    hnets: list[str] = []
    parse_hnet = False
    if line[0] != "[":
        nets.append("")
    for ch in line:
        if ch == "[":
            parse_hnet = True
            hnets.append("")
        elif ch == "]":
            parse_hnet = False
            nets.append("")
        elif parse_hnet:
            hnets[-1] += ch
        else:
            nets[-1] += ch
    return (nets, hnets)


def main() -> None:
    with open("day07.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2016/day07-1.ts
type IPV7 = {
  nets: string[];
  hnets: string[];
};

function solution(content: string) {
  return content
    .split("\n")
    .map((line) => parse(line.trim()))
    .filter(supportTLS).length;
}

function supportTLS({ nets, hnets }: IPV7): boolean {
  return nets.some(hasABBA) && !hnets.some(hasABBA);
}

function hasABBA(part: string): boolean {
  if (part.length < 4) return false;
  return part
    .split("")
    .some(
      (p0, index, array) =>
        p0 === array[index + 3] &&
        array[index + 1] === array[index + 2] &&
        p0 !== array[index + 1]
    );
}

function parse(line: string): IPV7 {
  const nets: string[] = [];
  const hnets: string[] = [];
  let parseHnet: boolean = false;
  if (line[0] !== "[") nets.push("");
  line.split("").forEach((char) => {
    if (char === "[") {
      parseHnet = true;
      hnets.push("");
    } else if (char === "]") {
      parseHnet = false;
      nets.push("");
    } else if (parseHnet) {
      hnets[hnets.length - 1] += char;
    } else {
      nets[nets.length - 1] += char;
    }
  });
  return {
    nets,
    hnets,
  };
}

Deno.readTextFile("day07.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
